source/common/ssl/context_config_impl.cc

```diff
@@ -34,6 +34,61 @@ const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES =
 
 const std::string ContextConfigImpl::DEFAULT_ECDH_CURVES = "X25519:P-256";
 
+static std::vector<bssl::UniquePtr<X509_CRL>>
+loadCRL(const envoy::api::v2::CertificateValidationContext& vctx) {
+  std::vector<bssl::UniquePtr<X509_CRL>> crls;
+
+  if (!vctx.has_crl()) {
+    return crls;
+  }
+
+  // We want to load multiple CRLs from the same file, so we need to use
+  // OpenSSL's STACK abstraction. We read from the input file or inline data,
+  // and load into the stack of X509_INFOs.
+  bssl::UniquePtr<STACK_OF(X509_INFO)> xis(nullptr);
+
+  auto crl_conf = vctx.crl();
+  if (crl_conf.filename().length()) {
+    std::unique_ptr<FILE, decltype(&fclose)> fp(fopen(crl_conf.filename().c_str(), "r"), &fclose);
+
+    if (!fp.get()) {
+      throw EnvoyException(
+          fmt::format("Failed to open CRL file '{}'", crl_conf.filename().c_str()));
+    }
+
+    xis.reset(PEM_X509_INFO_read(fp.get(), nullptr, nullptr, nullptr));
+  } else {
+    std::unique_ptr<BIO, decltype(&BIO_free_all)> bio(
+        BIO_new_mem_buf(crl_conf.inline_().c_str(), crl_conf.inline_().length()), BIO_free_all);
+
+    if (!bio.get()) {
+      throw EnvoyException(fmt::format("Failed to load inline CRL"));
+    }
+
+    xis.reset(PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));
+  }
+
+  if (!xis) {
+    throw EnvoyException(fmt::format("Failed to parse CRL"));
+  }
+
+  // Iterate through every item in the stack, adding all found CRLs to our
+  // return array.
+  for (unsigned int i = 0; i < sk_X509_INFO_num(xis.get()); i++) {
+    auto xi = sk_X509_INFO_value(xis.get(), i);
+    if (xi->crl != nullptr) {
+      crls.emplace_back(xi->crl);
+
+      // We need to set the pointer to nullptr here, since otherwise the
+      // pointer's destructor (`sk_X509_INFO_pop_free`) will free the
+      // underlying CRL, that we've just copied to the vector.
+      xi->crl = nullptr;
```

Comment:
You should use `X509_CRL_up_ref()` instead of this hack.

Reply:
Thanks for the tip - switched to that!